### PR TITLE
Change verify_SSL default to 1, add ENV var to enable insecure default [CVE-2023-31486]

### PR DESCRIPTION
**Pull request by stigtsp**
To address #152

- Changes the `verify_SSL` default parameter from `0` to `1`

  Based on patch by Dominic Hargreaves:
  https://salsa.debian.org/perl-team/interpreter/perl/-/commit/1490431e40e22052f75a0b3449f1f53cbd27ba92

  Fixes CVE-2023-31486

- Add check for `$ENV{PERL_HTTP_TINY_SSL_INSECURE_BY_DEFAULT}` that
  enables the previous insecure default behaviour if set to `1`.

  This provides a workaround for users who encounter problems with the
  new `verify_SSL` default.

  Example to disable certificate checks:
  ```
    $ PERL_HTTP_TINY_SSL_INSECURE_BY_DEFAULT=1 ./script.pl
  ```

- Updates to documentation:
  - Describe changing the verify_SSL value
  - Describe the escape-hatch environment variable
  - Remove rationale for not enabling verify_SSL
  - Add missing certificate search paths
  - Replace "SSL" with "TLS/SSL" where appropriate
  - Use "machine-in-the-middle" instead of "man-in-the-middle"

- Update `210_live_ssl.t`
  - Use github.com, cpan.org and badssl.com hosts for checking
    certificates.
  - Add self signed snake-oil certificate for checking failures rather
    than bypassing the `SSL_verify_callback`
  - Test `verify_SSL` parameter in addition to low level SSL_options
  - Test that `PERL_HTTP_TINY_SSL_INSECURE_BY_DEFAULT=1` behaves as
    expected against badssl.com

- Added `180_verify_SSL.t`
  - Test that `verify_SSL` default is `1`
  - Test that `PERL_HTTP_TINY_SSL_INSECURE_BY_DEFAULT` behaves as expected
  - Test that using different values for `verify_SSL` and legacy `verify_ssl`
    doesn't disable cert checks

### CODE REVIEW

**Comment by haarg** [comment]
I don't love the name of the environment variable. If it's going to override the choice of whatever code is using HTTP::Tiny, rather than just changing the default, it should be more explicit about that. `PERL_HTTP_TINY_FORCE_INSECURE` or something. Or possibly it should only change the default, rather than overriding.

**Comment by nrdvana** [comment]
@haarg See [my comment](https://github.com/chansen/p5-http-tiny/issues/152#issuecomment-1555032209) from the related issue.  I feel it's fairly important for an environment variable to exist so that we provide an escape hatch for users surprised by these changes who suddenly discover they don't have a working trust store.

For comparison, LWP::UserAgent uses an environment variable ```PERL_LWP_SSL_VERIFY_HOSTNAME``` and Mojo::UserAgent uses a broader environment variable ```MOJO_INSECURE```.  I suggested ```HTTP_TINY_VERIFY_SSL```.  ```PERL_HTTP_TINY_INSECURE``` isn't bad though I prefer the more specific description of in what way it's insecure.

**Comment by stigtsp** [comment]
@haarg @nrdvana I agree that the environment variable name should be more descriptive, both to indicate that it forces behavior and that it applies to TLS/SSL.

What about `PERL_HTTP_TINY_FORCE_INSECURE_SSL=1`? (I'd like to use "TLS" instead, but maybe it would just be confusing).

**Comment by haarg** [comment]
> For comparison, LWP::UserAgent uses an environment variable `PERL_LWP_SSL_VERIFY_HOSTNAME` and Mojo::UserAgent uses a broader environment variable `MOJO_INSECURE`.

Both of these environment variables set defaults. They do not override an explicit setting by the user of the library.

**Review by rwp0 — APPROVED** [review]

**Comment by rjbs** [comment]
Having the environment variable affect the actual setting, overriding explicit requests from the user, seems odd.

I would have thought, instead, we'd say something like:

```perl
$default_verify
  = length $ENV{PERL_HTTP_TINY_VERIFY_CERTIFICATES} 
  ? !! $ENV{PERL_HTTP_TINY_VERIFY_CERTIFICATES}
  : 1;
```

**Comment by haarg** [comment]
I would definitely prefer that the environment variable only impact the default.

The only catch with that is that because HTTP::Tiny has been acting insecurely until now, many users are already overriding the default to be true. Many of these would probably be fine having an environment variable that could tell them to act insecurely.

Possibly this indicates that there should be two environment variables. I'm not entirely sure if that is necessary though. But if there is a variable that overrides the default, it definitely needs to be named appropriately to indicate that it is an override.

**Comment by nrdvana** [comment]
Another way of looking at it is that most frequently, a module author is making the decision of whether to verify SSL, but it is the end user who actually knows if they have a working trust store and the risks involved.  It seems right to me that the user's choice should override the module author's choice.

We're also mostly considering authors who add ```verify_SSL => 1```, but an author could also write ```verify_SSL => 0``` and what would that even mean?  That the author knows better than the user that some site should never be checked?  If I'm the user connecting to a self-signed certificate site, I'd rather add the cert to my trust store and go back to ```verify_SSL => 1``` via some environment variable.

Environment that only changes the default is a good pattern for most of the perl https clients, but only because most of them use a default of true and no module authors feel the need to override the default in normal usage.

**Comment by SoumyaWind** [comment]
Hi
I am looking for ([CVE-2023-31486](https://nvd.nist.gov/vuln/detail/CVE-2023-31486)) fix. Is this the right fix? Could this be merged anytime soon?

**Comment by stigtsp** [comment]
@rjbs I don't have a strong opinion about the environment variable, other that one should be included in the upstream patch to provide an escape hatch. Would you like me to change the PR to that behavior?

I do like `PERL_HTTP_TINY_FORCE_INSECURE_SSL=1` though, as it explicitly describes that insecure behavior is forced.

**Comment by stigtsp** [comment]
> Hi I am looking for ([CVE-2023-31486](https://nvd.nist.gov/vuln/detail/CVE-2023-31486)) fix. Is this the right fix? Could this be merged anytime soon?

While we wait for some agreement on the environment variable, and maybe other details, you can check out patches already included in some distributions:

- NixOS: https://github.com/NixOS/nixpkgs/pull/187480
- Alpine: https://gitlab.alpinelinux.org/alpine/aports/-/issues/14951
- Gentoo: https://bugs.gentoo.org/show_bug.cgi?id=905296

This PR and those patches are all based on [Debian Proposed Patch: \[PATCH\] Enable SSL by default in HTTP::Tiny](https://salsa.debian.org/perl-team/interpreter/perl/-/commit/1490431e40e22052f75a0b3449f1f53cbd27ba92.patch)

**Comment by rjbs** [comment]
@stigtsp I think changing the default is the correct behavior, and with not much weighing in elsewise, let's do that.  I think that `PERL_HTTP_TINY_FORCE_INSECURE_SSL` is not the right name if the default is changed, because things aren't forced to insecure behavior.  I would say `PERL_HTTP_TINY_SSL_INSECURE_BY_DEFAULT` gets the point across.

File: lib/HTTP/Tiny.pm [hunk 1]
```diff
@@ -40,8 +40,10 @@ This constructor returns a new HTTP::Tiny object.  Valid attributes include:
 * C<timeout> — Request timeout in seconds (default is 60) If a socket open,
   read or write takes longer than the timeout, the request response status code
   will be 599.
-* C<verify_SSL> — A boolean that indicates whether to validate the SSL
-  certificate of an C<https> — connection (default is false)
+* C<verify_SSL> — A boolean that indicates whether to validate the TLS/SSL
+  certificate of an C<https> — connection (default is true)
```
Comment (xdg) [inline lib/HTTP/Tiny.pm:44]: This and the environment variable below should note the version of HTTP::Tiny at which this default changed.  People might read current docs online but have older versions installed.

**Comment by rjbs** [comment]
Thanks very much @stigtsp, I am now moving on this.

File: lib/HTTP/Tiny.pm [hunk 2]
```diff
@@ -115,7 +117,7 @@ sub new {
         max_redirect => 5,
         timeout      => defined $args{timeout} ? $args{timeout} : 60,
         keep_alive   => 1,
-        verify_SSL   => $args{verify_SSL} || $args{verify_ssl} || 0, # no verification by default
+        verify_SSL   => $args{verify_SSL} // $args{verify_ssl} // _verify_SSL_default(),
```
Comment (xdg) [inline lib/HTTP/Tiny.pm:120]: Use of `//` bumps the minimum required Perl, which is currently v5.6.  I have no opinion about this change, but if we're going to make this change, this file should have `use v5.10.1` in the opening and `[MinimumPerl]` should be added to the `dist.ini` file.  Possibly `[OnlyCorePrereqs]` should be added as well with appropriate options to ensure HTTP::Tiny can continue to be dual-lifed.

File: lib/HTTP/Tiny.pm [hunk 3]
```diff
@@ -40,8 +40,10 @@ This constructor returns a new HTTP::Tiny object.  Valid attributes include:
 * C<timeout> — Request timeout in seconds (default is 60) If a socket open,
   read or write takes longer than the timeout, the request response status code
   will be 599.
-* C<verify_SSL> — A boolean that indicates whether to validate the SSL
-  certificate of an C<https> — connection (default is false)
+* C<verify_SSL> — A boolean that indicates whether to validate the TLS/SSL
+  certificate of an C<https> — connection (default is true)
+* C<$ENV{PERL_HTTP_TINY_SSL_INSECURE_BY_DEFAULT}> - Disables TLS/SSL certificate
+  verification by default if set to 1
```
Comment (xdg) [inline lib/HTTP/Tiny.pm:46]: This should clarify that this only applies if `verify_SSL` is not set.

File: lib/HTTP/Tiny.pm [hunk 4]
```diff
@@ -1744,52 +1753,39 @@ of L<IO::Socket::INET> for transparent support for both IPv4 and IPv6.
...
-Server identity verification is controversial and potentially tricky because it
-depends on a (usually paid) third-party Certificate Authority (CA) trust model
-to validate a certificate as legitimate.  This discriminates against servers
-with self-signed certificates or certificates signed by free, community-driven
-CA's such as L<CAcert.org|http://cacert.org>.
-
-By default, HTTP::Tiny does not make any assumptions about your trust model,
-threat level or risk tolerance.  It just aims to give you an encrypted channel
-when you need one.
+B<By default, HTTP::Tiny verifies server identity>.
```
Comment (xdg) [inline lib/HTTP/Tiny.pm:1774]: Please note here as well what version this default changes.

File: t/210_live_ssl.t [hunk 5]
```diff
@@ -27,33 +27,73 @@ if ( can_run('openssl') ) {
...
+    pass => { verify_SSL => 1 },
+    fail => { verify_SSL => 1, SSL_options => { SSL_ca_file => "t/snake-oil.crt" } },
+    default_should_yield => '1',
+});
+
+test_ssl('https://wrong.host.badssl.com/' => {
+    host => 'wrong.host.badssl.com',
+    pass => { SSL_options => { SSL_verifycn_scheme => 'none', SSL_verifycn_name => 'wrong.host.badssl.com', SSL_verify_mode => 0x00 } },
+    fail => { SSL_options => { SSL_verifycn_scheme => 'http', SSL_verifycn_name => 'wrong.host.badssl.com', SSL_verify_mode => 0x01, SSL_ca_file => Mozilla::CA::SSL_ca_file() } },
+    default_should_yield => '',
```
Comment (xdg) [inline t/210_live_ssl.t:48]: Nit: it's hard to know what this is doing or what the values should be.  Maybe call it `default_verify_should_return` and use `!!0` or `!!1` for all of them to help signal we're looking for Perl's true/false values?

**Review by xdg — CHANGES_REQUESTED** [review]
Looks good.  I have some doc/build comments and a readability nit.

File: lib/HTTP/Tiny.pm (hunk 2 above)
Comment (mauke) [inline lib/HTTP/Tiny.pm:120]: That's an interesting change in behavior. Consider:
```perl
my $http = HTTP::Tiny->new(
    verify_SSL => 0,
    verify_ssl => 1,
);
```
With the old code, this turns host verification on. With this change, this turns host verification **off**.

**Comment by stigtsp** [comment]
@xdg Thanks for the feedback! I've updated the commit:
- `//` has been avoided
- Behavior when using `verify_ssl` and `verify_SSL` attributes as pointed out by @mauke has been fixed, and tests for that has been added. 
- Docs are updated, and I've added a line at the top of the TLS section to describe the ENV var. Let me know if you need any changes to the text. I'm assuming version 0.083 is next btw.
- Your suggestions for `210_live_ssl.t` has been added

File: t/180_verify_SSL.t [hunk 6]
```diff
@@ -0,0 +1,108 @@
+#!perl
+
+use strict;
+use warnings;
+use Test::More 0.88;
+use lib 't';
+
+use HTTP::Tiny;
+
```
Comment (haarg) [inline t/180_verify_SSL.t:9]: This test should delete `PERL_HTTP_TINY_INSECURE_BY_DEFAULT` from the environment to start.

Comment (stigtsp) [reply t/180_verify_SSL.t:9]: Thanks, this is fixed in both `180_verify_SSL.t` and `210_live_ssl.t`

File: lib/HTTP/Tiny.pm (hunk 2 above)
Comment (stigtsp) [reply lib/HTTP/Tiny.pm]: Fix to avoid `//`

**Comment by akhuettel** [comment]
Question: by introducing the variable, don't you just introduce a fresh (allthough less critical) bypass vulnerability? 

(For similar reasons we're already building Perl with -DNO_PERL_RAND_SEED in Gentoo, and I can see us patching out support for this bypass here as well.)

**Comment by Grinnz** [comment]
verify_ssl can be turned off easily in code regardless (and this capability must be present to support local requests with self-signed certs). the only reason for the variable is as an option to keep code running without needing code changes. the insecurity here is the default, which is being changed

**Comment by akhuettel** [comment]
> verify_ssl can be turned off easily in code regardless (and this capability must be present to support local requests with self-signed certs). the only reason for the variable is as an option to keep code running without needing code changes. the insecurity here is the default, which is being changed

You might as well overlay a local HTTP::Tiny for such fragile code. Similar amount of ENV messing and noone else affected.

**Comment by nrdvana** [comment]
@akhuettel Gentoo didn't patch out ``-k`` in curl, or ``--no-check-certificate`` in wget, or ``PERL_LWP_SSL_VERIFY_HOSTNAME`` in libwww-perl.  Why would this one be any different?

**Review by xdg — APPROVED** [review]

**Comment by xdg** [comment]
HTTP-Tiny-0.083-TRIAL.tar.gz released to CPAN. Thanks everyone for contributing to this change!

**Comment by xdg** [comment]
HTTP-Tiny-0.084.tar.gz released to CPAN.

**Pull request merged**